pkg/mgmt/csipv/leak_protection.go

```diff
@@ -0,0 +1,359 @@
+	return nil
+}
+
+func (c *LeakProtectionController) removeFinalizer(pvc *corev1.PersistentVolumeClaim) error {
+	finalizer := c.GetFinalizer()
+	claimClone := pvc.DeepCopy()
+
+	// remove the annotation added previously.
+	delete(claimClone.ObjectMeta.Annotations, c.GetAnnotationKey())
+
+	currFinalizerList := claimClone.ObjectMeta.Finalizers
+	newFinalizerList := make([]string, 0, len(currFinalizerList))
+	for _, v := range currFinalizerList {
+		if v == finalizer {
+			continue
+		}
+		newFinalizerList = append(newFinalizerList, v)
+	}
+	claimClone.ObjectMeta.Finalizers = newFinalizerList
+
+	_, err := c.client.CoreV1().PersistentVolumeClaims(claimClone.Namespace).Update(context.TODO(), claimClone, metav1.UpdateOptions{})
+	if err != nil {
+		klog.ErrorS(err, "failed to remove finalizer from PVC",
+			"finalizer", finalizer,
+			"pvc", klog.KObj(pvc))
+		return err
+	}
+	klog.V(3).InfoS("removed finalizer from PVC",
+		"finalizer", finalizer, "pvc", klog.KObj(pvc))
+	return nil
+}
+
+// BeginCreateVolume add relevant finalizer to the given pvc to avoid potential
+// csi volume leak. It must be called from the create volume csi method
+// implementation just before actual volume provisioning.
+// volumeName param should be same as csi.CreateVolumeRequest Name parameter.
+// In case of error, the csi driver should return non-retryable grpc error codes
+// to external provisioner.
+// Returned finishCreateVolume function must be called (preferably under defer)
+// after attempting to provision volume.
+// e.g
+// {
+//		finishCreateVolume, err := c.BeginCreateVolume("volumeId", "namespace", "name")
+//		if err != nil {
+//			return nil, status.Errorf(codes.FailedPrecondition, err.Error())
+//		}
+//		defer finishCreateVolume()
+//		..... start provisioning volume here .....
+// }
+func (c *LeakProtectionController) BeginCreateVolume(volumeName,
+	pvcNamespace, pvcName string) (func(), error) {
+	pvc, err := c.client.CoreV1().PersistentVolumeClaims(pvcNamespace).
+		Get(context.TODO(), pvcName, metav1.GetOptions{})
+	if err != nil {
+		klog.ErrorS(err, "failed to fetch pvc", "pvc", klog.KRef(pvcNamespace, pvcName))
+		return nil, status.Errorf(codes.FailedPrecondition, "failed to fetch pvc: %v", err)
+	} else if pvc.GetDeletionTimestamp() != nil {
+		// if pvc is already marked for deletion, return err.
+		err = fmt.Errorf("pvc already marked for deletion")
+		klog.ErrorS(err, "", "pvc", klog.KRef(pvcNamespace, pvcName))
+		return nil, status.Errorf(codes.FailedPrecondition, err.Error())
+	}
+
+	key := c.claimsInProgressKey(pvc)
+	finishCreateVolume := func() {
+		c.claimsInProgress.Remove(key)
+	}
+	alreadyExists := c.claimsInProgress.Add(key)
+	if alreadyExists {
+		return nil, status.Errorf(codes.Aborted,
+			"csi driver already has volume creation in progress")
+	}
+
+	if err = c.addFinalizer(pvc, volumeName); err != nil {
+		finishCreateVolume() // make sure we clean up on error.
+		return nil, err
```

Comment:
can we return the func here also instead of calling it here explicitly and then return nil  ?

Reply:
It is cause of non-nil error. Essentially, the caller should only consider return value as valid if the error nil. Otherwise, the caller have to check if return value is valid every time regardless of error is nil or not.